### PR TITLE
[styled-system] Make N a generic on style and add typedef for scale

### PR DESCRIPTION
**Pull request by Dhalton**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
https://github.com/styled-system/styled-system/blob/master/src/index.js
https://styled-system.com/guides/theming
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This is a follow-up PR after some suggestions were made on #35265.

@StevenLangbroek this *should* fit your needs; I used your example in a new test.
One thing that I'm still not totally sure on is the typedef for `scale`, so if anyone has any insight on that, I'd appreciate it.

Other than that, I apologize for the wait but I hope this works for y'all! 😄

### CODE REVIEW

File: types/styled-system/index.d.ts [hunk 1]
```diff
@@ -36,17 +36,20 @@ export interface styleFn {
...
 }
 
-export interface LowLevelStylefunctionArguments {
-    prop: string;
-    cssProperty?: string;
-    alias?: string;
-    key?: string;
-    transformValue?: (n: string | number, scale: Array<string | number>) => any;
-    scale?: Array<string | number>;
+export type Scale = object | Array<string | number>;
```
Comment (alloy) [inline types/styled-system/index.d.ts]: Do those “nested objects” not follow a particular shape?

Comment (StevenLangbroek) [reply types/styled-system/index.d.ts]: Scales aren't known upfront, so should be generic below too actually. (we use deeply nested objects for example).

File: types/styled-system/index.d.ts [hunk 2]
```diff
@@ -36,17 +36,20 @@ export interface styleFn {
...
+
+export interface LowLevelStylefunctionArguments<N> {
+  prop: string;
+  cssProperty?: string;
+  alias?: string;
+  key?: string;
+  transformValue?: (n: N, scale?: Scale) => any;
```
Comment (StevenLangbroek) [inline types/styled-system/index.d.ts]: Can you make this generic on `LowLevelStylefunctionArguments<N, S = Scale>`?

Comment (StevenLangbroek) [reply types/styled-system/index.d.ts]: Test could be my previous example:

```ts
const buttonSizes = {
  sm: '8px',
  md: '12px',
  lg: '24px',
};

const buttonSizeWithGeneric = style<keyof buttonSizes, typeof buttonSizes>({
  prop: 'size',
  cssProperty: 'font-size',
  transformValue: (size, sizes) => sizes[size]
});
```

**Review by StevenLangbroek — CHANGES_REQUESTED** [review]
Thanks for taking the time to revisit this 🙇, requested 1 change.

**Review by peduarte — APPROVED** [review]

**Comment by alloy** [comment]
@StevenLangbroek If you feel strongly about your request then you should leave a GH review with the “request changes” status, otherwise a maintainer with commit access is going to fly by and accidentally merge it.

File: types/styled-system/index.d.ts (hunk 2 above)
Comment (Dhalton) [reply types/styled-system/index.d.ts]: Will do!

**Comment by StevenLangbroek** [comment]
@alloy I did.

File: types/styled-system/index.d.ts (hunk 2 above)
Comment (Dhalton) [reply types/styled-system/index.d.ts]: @StevenLangbroek I updated this PR to make scale a generic. I had to do it on `style`, the same way I did it for N, but it should work. Let me know if you're good with this change!

**Review by StevenLangbroek — APPROVED** [review]
Thanks so much! 💯🥂

**Comment by alloy** [comment]
Great, thanks for all the discussion 👌

**Pull request merged**